datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
     test_dataset = tf.data.Dataset.from_tensor_slices(test_images).shuffle(test_size).batch(batch_size)
     return train_dataset, test_dataset
 
+'''
+Spoken digit audio dataset
+'''
+
+def _preprocess_spoken_digit_audio_tf(example, with_labels):
+    audio = example['audio']
+    # convert from int16? to float32 [-1.0, 1.0]
+    if audio.dtype.is_integer:
+        # detected clipping using just tf.int16.max to normalize...
+        audio = tf.cast(audio, tf.float32) / (tf.int16.max + 1.0)
+        
+    # remove DC offset
+    audio_mean = tf.math.reduce_mean(audio, axis=0) 
+    audio -= audio_mean
```

Comment:
Data normalization should not be done with the mean value of the given example. Any kind of normalization should be done with the same parameters for all examples. These parameters could be either fixed or could come from the statistics of your dataset. Common cases:

- fixed: for positive-value data, normalize with max and then center around zero, i.e. x_norm = (x / x_max) - 0.5
- standardize: for any data, x_norm = (x - x_mean) / x_std, where mean and std are taken over your entire dataset.

We can go with the first one here which is simpler to calculate. I'd add a parameter to the function with a default value to represent x_max, say `max_value=2**16`.

Reply:
Again, though this kind of looks like the sort of standardization you're referring to, it isn't. What I'm doing here is removing the DC offset of the sample. When we have the entire sample available this is the easiest method of doing so and *must* be done using the mean of each individual sample. Using the mean of the entire dataset wouldn't accomplish the same goal, it would just shift the DC offset to some other value. 

If we were building a realtime system which works on short blocks of audio (rather than capturing a complete utterance / phrase), I would instead use a DC-blocking filter. But since we're performing inference on these clips which do contain the entire signal of interest, we can safely use this simpler approach. 

The reason I'm doing this stuff here is that when we load the .npy data we don't want to have to do it over again. It's not discarding any useful information or making any unreasonable assumptions about the nature of the input data. We just need floating point audio with minimal DC offset in the range (-1, 1), and that's not what we get from the raw TF dataset so I'm converting it here before we write to the .npy files so we don't have to worry about doing it later. 

The stuff you're talking about might apply to some of the features we later compute from the raw audio, but that's at a different point in the pipeline and so far the features we're considering are already constrained to reasonable ranges.